medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,595 @@
+        the edges of the patient, and removes patients with less than two unique time
+        windows.
+
+        For example, if a patient has absolute time windows [0, 2, 2, 3], the method
+        will assign relative time windows [0, 1, 1, 2] to the corresponding edges.
+
+        Args:
+            medrecord (MedRecord): The MedRecord to be modified.
+            patient_index (NodeIndex): The index of the patient being processed.
+            time_windows (List[MedRecordValue]): The list of absolute time windows for
+                the patient.
+            absolute_time_window_attribute (MedRecordAttribute): The attribute name for
+                absolute time windows.
+            time_window_attribute (MedRecordAttribute): The attribute name for relative
+                time windows.
+            number_windows_attribute (MedRecordAttribute): The attribute name for
+                the number of unique time windows.
+
+        Returns:
+            None: This method modifies the MedRecord in place and doesn't return
+                anything.
+
+        Note:
+            This method will remove the patient node if there are less than two unique
+                time windows.
+        """
+        unique_sorted = sorted(
+            set(time_windows), key=lambda x: (isinstance(x, bool), x)
+        )
+        value_to_rank = {value: rank for rank, value in enumerate(unique_sorted)}
+
+        # Remove patients with only one time window
+        if len(value_to_rank) < 2:
+            medrecord.remove_node(patient_index)
+            return
+
+        medrecord.node[patient_index, number_of_windows_attribute] = len(value_to_rank)
+
+        for single_edge in medrecord.edges_connecting(
+            patient_index,
+            sorted(medrecord.nodes_in_group(self.concepts_group)),
+            directed=False,
+        ):
+            medrecord.edge[single_edge, time_window_attribute] = value_to_rank[
+                medrecord.edge[single_edge, absolute_time_window_attribute]
+            ]
+
+    def _find_relative_times(
+        self,
+        medrecord: MedRecord,
+        first_admission_attribute: MedRecordAttribute,
+        concept_index_attribute: MedRecordAttribute,
+        time_interval_days: int,
+        min_codes_per_window: int,
+    ) -> Tuple[MedRecord, PreprocessingAttributes]:
+        """Finds the relative time windows for each patient and removes the time windows with less than min_codes_per_window codes.
+
+        For instance, consider a single patient, 'p1', with admission dates as follows:
+        'p1': ['2021-01-01', '2021-01-15', '2021-01-25']
+
+        If the `time_interval_days` is 7 days, the method will compute time window as
+        the floor division of the difference in days from the first admission and the
+        `time_interval_days`. So, the resulting ABSOLUTE time windows will be:
+        'p1': [0, 2, 2]
```

Comment:
wouldn't it be `'p1': [0, 2, 3]`?

Reply:
you are right! Nicely found! Good understanding!